kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/InternalCompletableFuture.java

```diff
@@ -73,9 +77,242 @@ public CompletionStage<T> minimalCompletionStage() {
+    public static <U> CompletableFuture<U> completedFuture(EventLoop executor, U value) {
         var f = new InternalCompletableFuture<U>(executor);
         f.complete(value);
         return f;
     }
+
+    @Override
+    public <U> CompletableFuture<U> thenApply(Function<? super T, ? extends U> fn) {
+        return super.thenCompose(t -> {
+            if (isInEventLoop()) {
+                return super.thenApply(fn);
+            }
+            else {
+                return super.thenApplyAsync(fn);
+            }
+        });
+    }
+
+    private boolean isInEventLoop() {
+        return eventLoop.inEventLoop();
+    }
+
+    @Override
+    public CompletableFuture<Void> thenAccept(Consumer<? super T> action) {
+        return super.thenCompose(t -> {
+            if (isInEventLoop()) {
+                return super.thenAccept(action);
+            }
+            else {
+                return super.thenAcceptAsync(action);
+            }
+        });
+    }
+
+    @Override
+    public CompletableFuture<Void> thenRun(Runnable action) {
+        return super.thenCompose(t -> {
+            if (isInEventLoop()) {
+                return super.thenRun(action);
+            }
+            else {
+                return super.thenRunAsync(action);
+            }
+        });
+    }
+
+    @Override
+    public <U, V> CompletableFuture<V> thenCombine(CompletionStage<? extends U> other, BiFunction<? super T, ? super U, ? extends V> fn) {
+        CompletionStage<? extends U> internalFuture = toInternalFuture(other);
+        return super.thenCombineAsync(internalFuture, fn);
+    }
+
+    private <U> CompletionStage<? extends U> toInternalFuture(CompletionStage<? extends U> other) {
+        if (other instanceof InternalCompletableFuture) {
+            return other;
+        }
+        CompletableFuture<U> incompleteFuture = newIncompleteFuture();
+        other.whenComplete((u, throwable) -> {
+            if (eventLoop.inEventLoop()) {
+                if (throwable != null) {
+                    incompleteFuture.completeExceptionally(throwable);
+                }
+                else {
+                    incompleteFuture.complete(u);
+                }
+            }
+            else {
+                eventLoop.execute(() -> {
+                    if (throwable != null) {
+                        incompleteFuture.completeExceptionally(throwable);
+                    }
+                    else {
+                        incompleteFuture.complete(u);
+                    }
+                });
+            }
+        });
+        return incompleteFuture;
+    }
+
+    @Override
+    public <U> CompletableFuture<Void> thenAcceptBoth(CompletionStage<? extends U> other, BiConsumer<? super T, ? super U> action) {
+        CompletionStage<? extends U> internalFuture = toInternalFuture(other);
+        // go async to ensure function executed on event loop
+        return super.thenAcceptBothAsync(internalFuture, action);
```

Comment:
there's a scheduling cost here to go async (a task will be added to a queue somewhere), but I don't think we've used any of these stage composition methods (*Either, *Both, thenCombine) in practise yet.

maybe you could craft something for each composition using other methods, but this feels sufficient